### PR TITLE
chore: unite the bin cache dir for all projects in go.work

### LINKED ISSUE #426: Should unite the binary cache directory
### Description

Currently, each module has its own binary cache, and some tools are cached in different directories if they are required in different Makefiles. Let's unite the binary cache path.

### Environment

- HTNN version:

### PR DESCRIPTION
**Pull request by Chever-John**
Fix #426 

I am working on issue #426.
I hope that my work is on the right track. I have thoroughly run every makefile command related to the project to ensure everything is functioning correctly.

Looking forward to your feedback and hoping my contribution is valuable.

Thank you!

### CODE REVIEW

File: controller/Makefile [hunk 1]
```diff
@@ -89,7 +89,7 @@ KIND ?= kind
 ##@ Build Dependencies
 
 ## Location to install dependencies to
-LOCALBIN ?= $(shell pwd)/bin
+LOCALBIN ?= $(shell cd .. && pwd)/bin
```
Comment (spacewander) [inline controller/Makefile:92]: What about using
```
LOCALBIN ?= $(dir $(lastword $(MAKEFILE_LIST)))bin
```

in the common.mk?

File: e2e/istio.sh [hunk 2]
```diff
@@ -17,7 +17,9 @@
 set -eo pipefail
 set -x
 
-HELM="$(pwd)/bin/helm"
+PARENT_DIR="$(cd .. && pwd)"
```
Comment (spacewander) [inline e2e/istio.sh:20]: We can pass LOCALBIN to the script, like https://github.com/mosn/htnn/blob/e95743bbc1dcb8355619780421ac93a9a787ad1b/Makefile#L44

**Review by spacewander — COMMENTED** [review]
Don't forget to update
e2e/Makefile
binary cache in .github/workflows/test.yml & .github/workflows/lint.yml

File: controller/Makefile (hunk 1 above)
Comment (spacewander) [reply controller/Makefile:92]: @Chever-John 
Would you give this a try? Thanks!

Comment (Chever-John) [reply controller/Makefile:92]: Sry for late~

Yes, I have tried it.

I added the following code (A) into the common.mk file and commented out the corresponding code in the controller/Makefile as shown below (B):

A. code
```makefile
LOCALBIN ?= $(dir $(lastword $(MAKEFILE_LIST)))
$(LOCALBIN):
	@mkdir -p $(LOCALBIN)
```

B. code

comment out the `LOCALBIN ?= $(shell cd .. && pwd)/bin`

```makefile
## Location to install dependencies to
#LOCALBIN ?= $(shell cd .. && pwd)/bin
# Remember to remove tools downloaded into bin directory manually before updating them.
# If they need to be updated frequently, we can consider to store them in the `Dockerfile.dev`.
$(LOCALBIN):
	mkdir -p $(LOCALBIN)
```

And I tried to run the command `make sth` in the `controller` dir as follows:

```shell
╭─cheverjohn at Jonathan’s MacBook Pro in ~/Workspace/OpenSource/github.com/Chever-John/htnngit:(chore/unite-the-bin-cache-dir*)
╰─± cd controller 
╭─cheverjohn at Jonathan’s MacBook Pro in ~/Workspace/OpenSource/github.com/Chever-John/htnn/controllergit:(chore/unite-the-bin-cache-dir*)
╰─± make controller-gen
test -s ..//controller-gen && ..//controller-gen --version | grep -q v0.13.0 || \
        GOBIN=../ go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.13.0
cannot install, GOBIN must be an absolute path
make: *** [..//controller-gen] Error 1
```

Then errors came out:)

Comment (Chever-John) [reply controller/Makefile:92]: How about use the following code:

```makefile
# common.mk
ROOT_DIR := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
LOCALBIN := $(ROOT_DIR)/bin
$(LOCALBIN):
	@mkdir -p $(LOCALBIN)
```

**Review by Chever-John — COMMENTED** [review]
@spacewander 

Here is my solution.

File: common.mk [hunk 3]
```diff
@@ -16,6 +16,11 @@ SHELL = /bin/bash
 OS = $(shell uname)
 IN_CI ?=
 
+ROOT_DIR := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
```
Comment (Chever-John) [inline common.mk:19]: linke to https://github.com/mosn/htnn/pull/609#discussion_r1660050608

Here is my solution:)

@spacewander

File: controller/Makefile [hunk 4]
```diff
@@ -89,11 +89,11 @@ KIND ?= kind
 ##@ Build Dependencies
 
 ## Location to install dependencies to
-LOCALBIN ?= $(shell cd .. && pwd)/bin
+#LOCALBIN ?= $(shell cd .. && pwd)/bin
 # Remember to remove tools downloaded into bin directory manually before updating them.
 # If they need to be updated frequently, we can consider to store them in the `Dockerfile.dev`.
-$(LOCALBIN):
-	mkdir -p $(LOCALBIN)
+#$(LOCALBIN):
```
Comment (Chever-John) [inline controller/Makefile:95]: Just comment out the above code temporarily.

**Review by Chever-John — COMMENTED** [review]
> Don't forget to update e2e/Makefile binary cache in .github/workflows/test.yml & .github/workflows/lint.yml

I found some code related to binary cache in the test.yml file. Is there any related code in the lint.yml file (perhaps not)?

File: controller/Makefile (hunk 1 above)
Comment (spacewander) [reply controller/Makefile]: @Chever-John 
LGTM

**Comment by spacewander** [comment]
> > Don't forget to update e2e/Makefile binary cache in .github/workflows/test.yml & .github/workflows/lint.yml
> 
> I found some code related to binary cache in the test.yml file. Is there any related code in the lint.yml file (perhaps not)?

You are right, there is no need to update lint.yml because it's already using `./bin`.

File: controller/Makefile (hunk 1 above)
Comment (Chever-John) [reply controller/Makefile]: Update!!!

from `ROOT_DIR := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))` to `ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))`

**Comment by Chever-John** [comment]
Everything is set up:)
I hope this will run smoothly.

**Comment by Chever-John** [comment]
During local testing, I encountered an issue where the local Go version is `1.22.x` and the controller-gen version is `0.13.0` (default in the controller project). When I execute the command `make manifests`, the following error occurs:

```shell
~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
☹  ../bin/controller-gen --version                                                                                     chore/unite-the-bin-cache-dir 53e855c
Version: v0.13.0

~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
☹  go version                                                                                                          chore/unite-the-bin-cache-dir 53e855c
go version go1.22.1 darwin/arm64

~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
☺  make manifests                                                                                                      chore/unite-the-bin-cache-dir 53e855c
/Users/cheverjohn/Workspace/OpenSource/github.com/Chever-John/htnn/bin/controller-gen rbac:roleName=htnn-role crd:ignoreUnexportedFields=true \
                output:crd:artifacts:config=../manifests/charts/htnn-controller/templates/crds/ \
                output:rbac:artifacts:config=../manifests/charts/htnn-controller/templates/rbac/ \
                paths="../types/..." paths="./internal/controller/..."
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102a1c604]

goroutine 182 [running]:
go/types.(*Checker).handleBailout(0x14000c27800, 0x14001811d18)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:367 +0x9c
panic({0x102cf5a60?, 0x103294af0?})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/runtime/panic.go:770 +0x124
go/types.(*StdSizes).Sizeof(0x0, {0x102dc0990, 0x10329d300})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:228 +0x314
go/types.(*Config).sizeof(...)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:333
go/types.representableConst.func1({0x102dc0990?, 0x10329d300?})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:76 +0x9c
go/types.representableConst({0x102dc6c30, 0x103268240}, 0x14000c27800, 0x10329d300, 0x14001811468)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:92 +0x138
go/types.(*Checker).representation(0x14000c27800, 0x14001814d00, 0x10329d300)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:256 +0x68
go/types.(*Checker).implicitTypeAndValue(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/expr.go:375 +0x304
go/types.(*Checker).assignment(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0}, {0x102b34081, 0x14})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:52 +0x23c
go/types.(*Checker).initConst(0x14000c27800, 0x14001dbf500, 0x14001814d00)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:126 +0x274
go/types.(*Checker).constDecl(0x14000c27800, 0x14001dbf500, {0x102dc3538, 0x1400132a1a0}, {0x102dc3538, 0x1400132a1c0}, 0x0)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:490 +0x250
go/types.(*Checker).objDecl(0x14000c27800, {0x102dcbd20, 0x14001dbf500}, 0x0)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:191 +0x84c
go/types.(*Checker).packageObjects(0x14000c27800)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/resolver.go:693 +0x468
go/types.(*Checker).checkFiles(0x14000c27800, {0x14001180000, 0x5, 0x5})
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:408 +0x164
go/types.(*Checker).Files(...)
        /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:372
sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck(0x1400025f4d0, 0x14001090fc0)
        /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:286 +0x2d8
sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo(0x14001090fc0)
        /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:99 +0x44
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0x140017a7290, 0x14001090fc0)
        /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:268 +0x304
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x5b?)
        /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:262 +0x58
created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check in goroutine 201
        /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:260 +0x230
make: *** [manifests] Error 2
```

This problem can be stably reproduced.

Related discussion as following issues: https://github.com/kubernetes-sigs/controller-tools/issues/888

I don't know how to fix this problem. Should I create a new issue to fix it or just solve it in this PR?

**Comment by spacewander** [comment]
> During local testing, I encountered an issue where the local Go version is `1.22.x` and the controller-gen version is `0.13.0` (default in the controller project). When I execute the command `make manifests`, the following error occurs:
> 
> ```shell
> ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> ☹  ../bin/controller-gen --version                                                                                     chore/unite-the-bin-cache-dir 53e855c
> Version: v0.13.0
> 
> ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> ☹  go version                                                                                                          chore/unite-the-bin-cache-dir 53e855c
> go version go1.22.1 darwin/arm64
> 
> ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> ☺  make manifests                                                                                                      chore/unite-the-bin-cache-dir 53e855c
> /Users/cheverjohn/Workspace/OpenSource/github.com/Chever-John/htnn/bin/controller-gen rbac:roleName=htnn-role crd:ignoreUnexportedFields=true \
>                 output:crd:artifacts:config=../manifests/charts/htnn-controller/templates/crds/ \
>                 output:rbac:artifacts:config=../manifests/charts/htnn-controller/templates/rbac/ \
>                 paths="../types/..." paths="./internal/controller/..."
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102a1c604]
> 
> goroutine 182 [running]:
> go/types.(*Checker).handleBailout(0x14000c27800, 0x14001811d18)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:367 +0x9c
> panic({0x102cf5a60?, 0x103294af0?})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/runtime/panic.go:770 +0x124
> go/types.(*StdSizes).Sizeof(0x0, {0x102dc0990, 0x10329d300})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:228 +0x314
> go/types.(*Config).sizeof(...)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:333
> go/types.representableConst.func1({0x102dc0990?, 0x10329d300?})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:76 +0x9c
> go/types.representableConst({0x102dc6c30, 0x103268240}, 0x14000c27800, 0x10329d300, 0x14001811468)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:92 +0x138
> go/types.(*Checker).representation(0x14000c27800, 0x14001814d00, 0x10329d300)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:256 +0x68
> go/types.(*Checker).implicitTypeAndValue(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/expr.go:375 +0x304
> go/types.(*Checker).assignment(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0}, {0x102b34081, 0x14})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:52 +0x23c
> go/types.(*Checker).initConst(0x14000c27800, 0x14001dbf500, 0x14001814d00)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:126 +0x274
> go/types.(*Checker).constDecl(0x14000c27800, 0x14001dbf500, {0x102dc3538, 0x1400132a1a0}, {0x102dc3538, 0x1400132a1c0}, 0x0)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:490 +0x250
> go/types.(*Checker).objDecl(0x14000c27800, {0x102dcbd20, 0x14001dbf500}, 0x0)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:191 +0x84c
> go/types.(*Checker).packageObjects(0x14000c27800)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/resolver.go:693 +0x468
> go/types.(*Checker).checkFiles(0x14000c27800, {0x14001180000, 0x5, 0x5})
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:408 +0x164
> go/types.(*Checker).Files(...)
>         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:372
> sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck(0x1400025f4d0, 0x14001090fc0)
>         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:286 +0x2d8
> sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo(0x14001090fc0)
>         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:99 +0x44
> sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0x140017a7290, 0x14001090fc0)
>         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:268 +0x304
> sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x5b?)
>         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:262 +0x58
> created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check in goroutine 201
>         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:260 +0x230
> make: *** [manifests] Error 2
> ```
> 
> This problem can be stably reproduced.
> 
> Related discussion as following issues: [kubernetes-sigs/controller-tools#888](https://github.com/kubernetes-sigs/controller-tools/issues/888)
> 
> Solution: change `controller-gen` version from v0.13.0 to v0.14.0.
> 
> I don't know how to fix this problem in which way. Should I create a new issue to fix it or just solve it in this PR?

Let's fix it in the new PR, Thanks!

**Review by spacewander — APPROVED** [review]

**Comment by Chever-John** [comment]
> > During local testing, I encountered an issue where the local Go version is `1.22.x` and the controller-gen version is `0.13.0` (default in the controller project). When I execute the command `make manifests`, the following error occurs:
> > ```shell
> > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > ☹  ../bin/controller-gen --version                                                                                     chore/unite-the-bin-cache-dir 53e855c
> > Version: v0.13.0
> > 
> > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > ☹  go version                                                                                                          chore/unite-the-bin-cache-dir 53e855c
> > go version go1.22.1 darwin/arm64
> > 
> > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > ☺  make manifests                                                                                                      chore/unite-the-bin-cache-dir 53e855c
> > /Users/cheverjohn/Workspace/OpenSource/github.com/Chever-John/htnn/bin/controller-gen rbac:roleName=htnn-role crd:ignoreUnexportedFields=true \
> >                 output:crd:artifacts:config=../manifests/charts/htnn-controller/templates/crds/ \
> >                 output:rbac:artifacts:config=../manifests/charts/htnn-controller/templates/rbac/ \
> >                 paths="../types/..." paths="./internal/controller/..."
> > panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> >         panic: runtime error: invalid memory address or nil pointer dereference
> > [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102a1c604]
> > 
> > goroutine 182 [running]:
> > go/types.(*Checker).handleBailout(0x14000c27800, 0x14001811d18)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:367 +0x9c
> > panic({0x102cf5a60?, 0x103294af0?})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/runtime/panic.go:770 +0x124
> > go/types.(*StdSizes).Sizeof(0x0, {0x102dc0990, 0x10329d300})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:228 +0x314
> > go/types.(*Config).sizeof(...)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:333
> > go/types.representableConst.func1({0x102dc0990?, 0x10329d300?})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:76 +0x9c
> > go/types.representableConst({0x102dc6c30, 0x103268240}, 0x14000c27800, 0x10329d300, 0x14001811468)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:92 +0x138
> > go/types.(*Checker).representation(0x14000c27800, 0x14001814d00, 0x10329d300)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:256 +0x68
> > go/types.(*Checker).implicitTypeAndValue(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/expr.go:375 +0x304
> > go/types.(*Checker).assignment(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0}, {0x102b34081, 0x14})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:52 +0x23c
> > go/types.(*Checker).initConst(0x14000c27800, 0x14001dbf500, 0x14001814d00)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:126 +0x274
> > go/types.(*Checker).constDecl(0x14000c27800, 0x14001dbf500, {0x102dc3538, 0x1400132a1a0}, {0x102dc3538, 0x1400132a1c0}, 0x0)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:490 +0x250
> > go/types.(*Checker).objDecl(0x14000c27800, {0x102dcbd20, 0x14001dbf500}, 0x0)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:191 +0x84c
> > go/types.(*Checker).packageObjects(0x14000c27800)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/resolver.go:693 +0x468
> > go/types.(*Checker).checkFiles(0x14000c27800, {0x14001180000, 0x5, 0x5})
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:408 +0x164
> > go/types.(*Checker).Files(...)
> >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:372
> > sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck(0x1400025f4d0, 0x14001090fc0)
> >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:286 +0x2d8
> > sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo(0x14001090fc0)
> >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:99 +0x44
> > sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0x140017a7290, 0x14001090fc0)
> >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:268 +0x304
> > sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x5b?)
> >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:262 +0x58
> > created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check in goroutine 201
> >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:260 +0x230
> > make: *** [manifests] Error 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This problem can be stably reproduced.
> > Related discussion as following issues: [kubernetes-sigs/controller-tools#888](https://github.com/kubernetes-sigs/controller-tools/issues/888)
> > Solution: change `controller-gen` version from v0.13.0 to v0.14.0.
> > I don't know how to fix this problem in which way. Should I create a new issue to fix it or just solve it in this PR?
> 
> Let's fix it in the new PR, Thanks!

So I'm curious, what kind of solution should I choose?

Do you choose to upgrade the version of `controller-gen` based on my suggestion?

**Review by spacewander — COMMENTED** [review]
Need to deal with
https://github.com/mosn/htnn/blob/efa91d02ccbfac51ff574deea6d86b3a1460b3a1/Makefile#L20
https://github.com/mosn/htnn/blob/efa91d02ccbfac51ff574deea6d86b3a1460b3a1/site/Makefile#L20

**Comment by spacewander** [comment]
> > > During local testing, I encountered an issue where the local Go version is `1.22.x` and the controller-gen version is `0.13.0` (default in the controller project). When I execute the command `make manifests`, the following error occurs:
> > > ```shell
> > > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > > ☹  ../bin/controller-gen --version                                                                                     chore/unite-the-bin-cache-dir 53e855c
> > > Version: v0.13.0
> > > 
> > > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > > ☹  go version                                                                                                          chore/unite-the-bin-cache-dir 53e855c
> > > go version go1.22.1 darwin/arm64
> > > 
> > > ~/Workspace/OpenSource/github.com/Chever-John/htnn/controller
> > > ☺  make manifests                                                                                                      chore/unite-the-bin-cache-dir 53e855c
> > > /Users/cheverjohn/Workspace/OpenSource/github.com/Chever-John/htnn/bin/controller-gen rbac:roleName=htnn-role crd:ignoreUnexportedFields=true \
> > >                 output:crd:artifacts:config=../manifests/charts/htnn-controller/templates/crds/ \
> > >                 output:rbac:artifacts:config=../manifests/charts/htnn-controller/templates/rbac/ \
> > >                 paths="../types/..." paths="./internal/controller/..."
> > > panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> > >         panic: runtime error: invalid memory address or nil pointer dereference
> > > [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102a1c604]
> > > 
> > > goroutine 182 [running]:
> > > go/types.(*Checker).handleBailout(0x14000c27800, 0x14001811d18)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:367 +0x9c
> > > panic({0x102cf5a60?, 0x103294af0?})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/runtime/panic.go:770 +0x124
> > > go/types.(*StdSizes).Sizeof(0x0, {0x102dc0990, 0x10329d300})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:228 +0x314
> > > go/types.(*Config).sizeof(...)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/sizes.go:333
> > > go/types.representableConst.func1({0x102dc0990?, 0x10329d300?})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:76 +0x9c
> > > go/types.representableConst({0x102dc6c30, 0x103268240}, 0x14000c27800, 0x10329d300, 0x14001811468)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:92 +0x138
> > > go/types.(*Checker).representation(0x14000c27800, 0x14001814d00, 0x10329d300)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/const.go:256 +0x68
> > > go/types.(*Checker).implicitTypeAndValue(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/expr.go:375 +0x304
> > > go/types.(*Checker).assignment(0x14000c27800, 0x14001814d00, {0x102dc09b8, 0x140002342a0}, {0x102b34081, 0x14})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:52 +0x23c
> > > go/types.(*Checker).initConst(0x14000c27800, 0x14001dbf500, 0x14001814d00)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/assignments.go:126 +0x274
> > > go/types.(*Checker).constDecl(0x14000c27800, 0x14001dbf500, {0x102dc3538, 0x1400132a1a0}, {0x102dc3538, 0x1400132a1c0}, 0x0)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:490 +0x250
> > > go/types.(*Checker).objDecl(0x14000c27800, {0x102dcbd20, 0x14001dbf500}, 0x0)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/decl.go:191 +0x84c
> > > go/types.(*Checker).packageObjects(0x14000c27800)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/resolver.go:693 +0x468
> > > go/types.(*Checker).checkFiles(0x14000c27800, {0x14001180000, 0x5, 0x5})
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:408 +0x164
> > > go/types.(*Checker).Files(...)
> > >         /Users/cheverjohn/Infra/dev-env/langs/go/go1.22.1/src/go/types/check.go:372
> > > sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck(0x1400025f4d0, 0x14001090fc0)
> > >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:286 +0x2d8
> > > sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo(0x14001090fc0)
> > >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/loader.go:99 +0x44
> > > sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0x140017a7290, 0x14001090fc0)
> > >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:268 +0x304
> > > sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x5b?)
> > >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:262 +0x58
> > > created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check in goroutine 201
> > >         /Users/cheverjohn/Workspace/golang/pkg/mod/sigs.k8s.io/controller-tools@v0.13.0/pkg/loader/refs.go:260 +0x230
> > > make: *** [manifests] Error 2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This problem can be stably reproduced.
> > > Related discussion as following issues: [kubernetes-sigs/controller-tools#888](https://github.com/kubernetes-sigs/controller-tools/issues/888)
> > > Solution: change `controller-gen` version from v0.13.0 to v0.14.0.
> > > I don't know how to fix this problem in which way. Should I create a new issue to fix it or just solve it in this PR?
> > 
> > 
> > Let's fix it in the new PR, Thanks!
> 
> So I'm curious, what kind of solution should I choose?
> 
> Do you choose to upgrade the version of `controller-gen` based on my suggestion?

Yes, let's update the version

File: e2e/istio.sh (hunk 2 above)
Comment (Chever-John) [reply e2e/istio.sh:20]: sry, I missed it here.

Please wait for me for a few minutes.

**Comment by Chever-John** [comment]
> Need to deal with
> 
> https://github.com/mosn/htnn/blob/efa91d02ccbfac51ff574deea6d86b3a1460b3a1/Makefile#L20
> 
> 
> https://github.com/mosn/htnn/blob/efa91d02ccbfac51ff574deea6d86b3a1460b3a1/site/Makefile#L20

Got.

Let me finish it right now:)

File: site/Makefile [hunk 5]
```diff
@@ -12,13 +12,9 @@
...
 
-SHELL = /bin/bash
+include ../common.mk
 
-LOCALBIN ?= $(shell pwd)/bin
-# Remember to remove tools downloaded into bin directory manually before updating them.
-# If they need to be updated frequently, we can consider to store them in the `Dockerfile.dev`.
-$(LOCALBIN):
-	@mkdir -p $(LOCALBIN)
+SHELL = /bin/bash
```
Comment (spacewander) [inline site/Makefile:17]: The common.mk already provides SHELL?

Comment (Chever-John) [reply site/Makefile:17]: Yes!

Have removed it!

**Comment by spacewander** [comment]
Merged. Thanks!

**Pull request merged**